// angular2-toaster/__ivy_ngcc__/src/toaster.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ToastComponent } from './toast.component';
import { ToasterContainerComponent } from './toaster-container.component';
import { ToasterService } from './toaster.service';
import * as ɵngcc0 from '@angular/core';
var ToasterModule = (function () {
    function ToasterModule() {
    }
ToasterModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ToasterModule });
ToasterModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ToasterModule_Factory(t) { return new (t || ToasterModule)(); }, providers: [ToasterService], imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ToasterModule, { declarations: function () { return [ToastComponent,
        ToasterContainerComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ToasterContainerComponent,
        ToastComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ToasterModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [
                    ToastComponent,
                    ToasterContainerComponent
                ],
                providers: [ToasterService],
                exports: [
                    ToasterContainerComponent,
                    ToastComponent
                ]
            }]
    }], function () { return []; }, null); })();
    return ToasterModule;
}());
export { ToasterModule };
/** @nocollapse */
ToasterModule.ctorParameters = function () { return []; };
