// ngx-treeview/__ivy_ngcc__/src/dropdown-menu.directive.js
import { Directive } from '@angular/core';
import { DropdownDirective } from './dropdown.directive';
import * as ɵngcc0 from '@angular/core';
var DropdownMenuDirective = /** @class */ (function () {
    function DropdownMenuDirective(dropdown) {
        this.dropdown = dropdown;
    }
    /** @nocollapse */
    DropdownMenuDirective.ctorParameters = function () { return [
        { type: DropdownDirective }
    ]; };
DropdownMenuDirective.ɵfac = function DropdownMenuDirective_Factory(t) { return new (t || DropdownMenuDirective)(ɵngcc0.ɵɵdirectiveInject(DropdownDirective)); };
DropdownMenuDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: DropdownMenuDirective, selectors: [["", "ngxDropdownMenu", ""]], hostVars: 4, hostBindings: function DropdownMenuDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("dropdown-menu", true)("show", ctx.dropdown.isOpen);
    } } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DropdownMenuDirective, [{
        type: Directive,
        args: [{
                selector: '[ngxDropdownMenu]',
                // tslint:disable-next-line:use-host-property-decorator
                host: {
                    '[class.dropdown-menu]': 'true',
                    '[class.show]': 'dropdown.isOpen'
                }
            }]
    }], function () { return [{ type: DropdownDirective }]; }, null); })();
    return DropdownMenuDirective;
}());
export { DropdownMenuDirective };
