// ngx-treeview/__ivy_ngcc__/src/treeview-config.js
import { Injectable } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
var TreeviewConfig = /** @class */ (function () {
    function TreeviewConfig() {
        this.hasAllCheckBox = true;
        this.hasFilter = false;
        this.hasCollapseExpand = false;
        this.decoupleChildFromParent = false;
        this.maxHeight = 500;
    }
    Object.defineProperty(TreeviewConfig.prototype, "hasDivider", {
        get: function () {
            return this.hasFilter || this.hasAllCheckBox || this.hasCollapseExpand;
        },
        enumerable: true,
        configurable: true
    });
    TreeviewConfig.create = function (fields) {
        var config = new TreeviewConfig();
        Object.assign(config, fields);
        return config;
    };
TreeviewConfig.ɵfac = function TreeviewConfig_Factory(t) { return new (t || TreeviewConfig)(); };
TreeviewConfig.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: TreeviewConfig, factory: function (t) { return TreeviewConfig.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(TreeviewConfig, [{
        type: Injectable
    }], function () { return []; }, null); })();
    return TreeviewConfig;
}());
export { TreeviewConfig };
